Lightning Lite docs (#10176)
PyTorchLightning/pytorch-lightning: docs/source/starter/lightning_lite.rst

The :class:`~pytorch_lightning.lite.LightningLite` provides you only with the tool to scale your training,
but there are several major challenges ahead of you now:

.. list-table::
   :widths: 50 50
   :header-rows: 0

   * - Processes divergence
     - This happens when processes execute different section of the code due to different if/else condition, race condition on existing files, etc., resulting in hanging.
   * - Cross processes reduction
     - Wrongly reported metrics or gradients due mis-reduction.
   * - Large sharded models
     - Instantiation, materialization and state management of large models.
   * - Rank 0 only actions
     - Logging, profiling, etc.
   * - Checkpointing / Early stopping / Callbacks
     - Ability to easily customize your training behaviour and make it stateful.
   * - Batch-level fault tolerance training
     - Ability to resume from a failure as if it never happened.

If you are facing one of those challenges then you are already meeting the limit of :class:`~pytorch_lightning.lite.LightningLite`.
We recommend you to convert to :doc:`Lightning <../starter/new-project>`, so you never have to worry about those.

----------

> Convert to Lightning

Reviewer: Hey @SkafteNicki. Good point, I would try to add maybe an extra step to demonstrate the convertion.

###########################################
LightningLite - Stepping Stone to Lightning
###########################################

.. image:: https://pl-public-data.s3.amazonaws.com/docs/static/images/lite/lightning_lite.gif
    :alt: Animation showing how to convert a standard training loop to a Lightning loop
    :width: 600px
    :align: center

|

:class:`~pytorch_lightning.lite.LightningLite` enables pure PyTorch users to scale their existing code
on any kind of device while retaining full control over their own loops and optimization logic.

:class:`~pytorch_lightning.lite.LightningLite` is the right tool for you if you match one of the two following descriptions:

- I want to quickly scale my existing code to multiple devices with minimal code changes.
- I would like to convert my existing code to the Lightning API, but a full path to Lightning transition might be too complex. I am looking for a stepping stone to ensure reproducibility during the transition.

> .. note:: LightningLite is a beta feature and will be considered stable for v1.6.

Reviewer 2: ```suggestion
.. warning:: `LightningLite` is currently a beta feature. API is subject to change.
```

Make this a warning. I wouldn't mention when this will be stable in case we need to change. We can still target this internally.

Reviewer: Done !

----------

****************
Learn by example
****************

My existing PyTorch code
========================

The ``run`` function contains custom training loop used to train ``MyModel`` on ``MyDataset`` for ``num_epochs`` epochs.

.. code-block:: python

    import torch
    from torch import nn
    from torch.utils.data import DataLoader, Dataset

    class MyModel(nn.Module):
        ...

    class MyDataset(Dataset):
        ...

    def run(args):

        device = "cuda" if torch.cuda.is_available() else "cpu"

        model = MyModel(...).to(device)
        optimizer = torch.optim.SGD(model.parameters(), ...)

        dataloader = DataLoader(MyDataset(...), ...)

        model.train()
        for epoch in range(args.num_epochs):
            for batch in dataloader:
                batch = batch.to(device)
                optimizer.zero_grad()
                loss = model(batch)
                loss.backward()
                optimizer.step()

    run(args)

----------

Convert to LightningLite
========================

Here are 5 required steps to convert to :class:`~pytorch_lightning.lite.LightningLite`.

> 1. Subclass :class:`~pytorch_lightning.lite.LightningLite` and override its :meth:`~pytorch_lightning.lite.LightningLite.run` method.
> 2. Move the body of your existing `run` function.
> 3. Remove all calls to `to`, `cuda` etc. as `LightningLite` will take care of it
> 4. Apply :meth:`~pytorch_lightning.lite.LightningLite.setup` over each model and optimizers pair, :meth:`~pytorch_lightning.lite.LightningLite.setup_dataloaders` on all your dataloaders and replace ``loss.backward()`` by ``self.backward(loss)``
> 5. Instantiate your :class:`~pytorch_lightning.lite.LightningLite` and call its :meth:`~pytorch_lightning.lite.LightningLite.run` method.

Reviewer 3: ```suggestion
1. Subclass :class:`~pytorch_lightning.lite.LightningLite` and override its :meth:`~pytorch_lightning.lite.LightningLite.run` method.
2. Move the body of your existing `run` function.
3. Remove all `.to`, `.cuda` etc calls since `LightningLite` will take care of it.
4. Apply :meth:`~pytorch_lightning.lite.LightningLite.setup` over each model and optimizers pair, :meth:`~pytorch_lightning.lite.LightningLite.setup_dataloaders` on all your dataloaders and replace ``loss.backward()`` by ``self.backward(loss)``.
5. Instantiate your :class:`~pytorch_lightning.lite.LightningLite` and call its :meth:`~pytorch_lightning.lite.LightningLite.run` method.
```

Reviewer: Done !

            self.broadcast(..., src=...)

            # The total number of processes running across all devices and nodes.
            self.world_size

            # The global index of the current process across all devices and nodes.
            self.global_rank

            # The index of the current process among the processes running on the local node.
            self.local_rank

            # The index of the current node.
            self.node_rank

            # Wether this global rank is rank zero.
            if self.is_global_zero:
                # do something on rank 0
                ...

            # Wait for all processes to enter this call.
            self.barrier()

    Lite(strategy="ddp", gpus=8, num_nodes=32, accelerator="gpu").run()

.. note:: We recommend instantiating the models within the :meth:`~pytorch_lightning.lite.LightningLite.run` method as large models would cause an out-of-memory error otherwise.

.. note::

    If you have hundreds or thousands of line within your :meth:`~pytorch_lightning.lite.LightningLite.run` function
    and you are feeling weird about it, this is exactly the expected feeling.

>     This is how our :class:`~pytorch_lightning.core.lightning.LightningModule` started 2-3 years ago

Reviewer 3: maybe mention the year since `2-3 years` is relative.
```suggestion
    This is how our :class:`~pytorch_lightning.core.lightning.LightningModule` started in 2019
```

Reviewer: Done !

    and then we started to organize the code for simplicity, interoperability and standardization.

----------

Distributed Training Pitfalls
=============================

The :class:`~pytorch_lightning.lite.LightningLite` provides you only with the tool to scale your training,
but there are several major challenges ahead of you now:

.. list-table::
   :widths: 50 50
   :header-rows: 0

>    * - Processes divergence
>      - This happens when processes execute different section of the code due to different if/else condition, race condition on existing files, etc., resulting in hanging.
>    * - Cross processes reduction
>      - Wrongly reported metrics or gradients due mis-reduction.
>    * - Large sharded models
>      - Instantiation, materialization and state management of large models.
>    * - Rank 0 only actions
>      - Logging, profiling, etc.
>    * - Checkpointing / Early stopping / Callbacks
>      - Ability to easily customize your training behaviour and make it stateful.
>    * - Batch-level fault tolerance training
>      - Ability to resume from a failure as if it never happened.

Reviewer 3: ```suggestion
   * - Processes divergence
     - This happens when processes execute a different section of the code due to different if/else conditions, race condition on existing files, etc., resulting in hanging.
   * - Cross processes reduction
     - Wrongly reported metrics or gradients due to mis-reduction.
   * - Large sharded models
     - Instantiation, materialization and state management of large models.
   * - Rank 0 only actions
     - Logging, profiling, etc.
   * - Checkpointing / Early stopping / Callbacks / Logging
     - Ability to easily customize your training behaviour and make it stateful.
   * - Batch-level fault tolerance training
     - Ability to resume from a failure as if it never happened.
```

Reviewer: Done !

If you are facing one of those challenges then you are already meeting the limit of :class:`~pytorch_lightning.lite.LightningLite`.
We recommend you to convert to :doc:`Lightning <../starter/new-project>`, so you never have to worry about those.

----------

Convert to Lightning
====================

The :class:`~pytorch_lightning.lite.LightningLite` is a stepping stone to transition fully to the Lightning API and benefits
from its hundreds of features.

You can see our :class:`~pytorch_lightning.lite.LightningLite` as a
future :class:`~pytorch_lightning.core.lightning.LightningModule` and slowly refactor / re-organize your code into its API.

If you have hundreds of line within your run function and feeling weird about it, it is exactly the expected feeling.

.. code-block:: python

    class Lite(LightningLite):

        # 1. This would becomes the LightningModule `__init__` function.

        def run(self, args):
            self.args = args

>             model = MyModel(...)
> 
>             self.fit(model)  # This would be automated by Lightning Trainer.

Reviewer 3: ```suggestion
            self.model = MyModel(...)

            self.fit()  # This would be automated by Lightning Trainer.
```

Reviewer: Done !

        # 2. This can be fully removed as Lightning handles the FitLoop
        # and setting up the model, optimizer, dataloader and many more.

        def fit(self):
            optimizer = self.configure_optimizers()

>             model, optimizer = self.setup(model, optimizer)
>             dataloader = self.setup_dataloaders(self.train_dataloader())
>             model.train()

Reviewer 3: ```suggestion
            self.model, optimizer = self.setup(self.model, optimizer)
            dataloader = self.setup_dataloaders(self.train_dataloader())
            self.model.train()
```

Reviewer: Done !

            for epoch in range(num_epochs):
                for batch in enumerate(dataloader):
                    optimizer.zero_grad()
                    loss = self.training_step(batch, batch_idx)
                    self.backward(loss)
                    optimizer.step()

        # 3. This stays here as it belongs to the LightningModule.

        def forward(self, x):
            return self.model(x)

        def training_step(self, batch, batch_idx):
            return self.forward(batch)

        def configure_optimizers(self):

>             return torch.optim.SGD(model.parameters(), ...)

Reviewer 3: ```suggestion
            return torch.optim.SGD(self.model.parameters(), ...)
```

Reviewer: Done !

    lite = Lite(strategy="ddp", accelerator="gpu", devices=4)

    # Running with the DDP Spawn strategy using 4 cpu processes
    lite = Lite(strategy="ddp_spawn", accelerator="cpu", devices=4)

Additionally, you can pass in your custom training type strategy by configuring additional parameters.

.. code-block:: python

    from pytorch_lightning.plugins import DeepSpeedPlugin

    lite = Lite(strategy=DeepSpeedPlugin(stage=2), accelerator="gpu", devices=2)

Support for Horovod and Fully Sharded training strategies are coming soon.

devices
=======

Configure the devices to run on. Can be of type:

- int: the number of devices (e.g., GPUs) to train on
- list of int: which device index (e.g., GPU ID) to train on (0-indexed)
- str: a string representation of one of the above

.. code-block:: python

    # default used by Lite, i.e., use the CPU
    lite = Lite(devices=None)

    # equivalent
    lite = Lite(devices=0)

    # int: run on 2 GPUs
    lite = Lite(devices=2, accelerator="gpu")

    # list: run on GPUs 1, 4 (by bus ordering)
    lite = Lite(devices=[1, 4], accelerator="gpu")
    lite = Lite(devices="1, 4", accelerator="gpu")  # equivalent

    # -1: run on all GPUs

>     lite = Lite(devices=-1)
>     lite = Lite(devices="-1")  # equivalent

Reviewer 3: ```suggestion
    lite = Lite(devices=-1, accelerator="gpu")
    lite = Lite(devices="-1", accelerator="gpu")  # equivalent
```
I think accelerator is still required.

Reviewer: Done !

    class Lite(LightningLite):
        def run(self):

            # Transfer and concatenate tensors across processes
            self.all_gather(...)

            # Transfer an object from one process to all the others
            self.broadcast(..., src=...)

            # The total number of processes running across all devices and nodes.
            self.world_size

            # The global index of the current process across all devices and nodes.
            self.global_rank

            # The index of the current process among the processes running on the local node.
            self.local_rank

            # The index of the current node.
            self.node_rank

            # Wether this global rank is rank zero.
            if self.is_global_zero:
                # do something on rank 0
                ...

            # Wait for all processes to enter this call.
            self.barrier()

    Lite(strategy="ddp", gpus=8, num_nodes=32, accelerator="gpu").run()

.. note:: We recommend instantiating the models within the :meth:`~pytorch_lightning.lite.LightningLite.run` method as large models would cause an out-of-memory error otherwise.

.. note:

>     If you require a custom data or model device placement, you can deactivate

Reviewer: @justusschock Mind checking this, it is a bit of a mouthful.

Reviewer: Done !

    Lite(strategy="ddp", gpus=8, num_nodes=32, accelerator="gpu").run()

If you require a custom data or model device placement, you can deactivate

> :class:`~pytorch_lightning.lite.LightningLite` automatic placement by doing

Reviewer 2: ```suggestion
:class:`~pytorch_lightning.lite.LightningLite` automatic device placement by doing
```
